compiler/src/main/java/com/squareup/anvil/compiler/codegen/reference/AnnotationReferenceIr.kt

```diff
@@ -0,0 +1,80 @@
+package com.squareup.anvil.compiler.codegen.reference
+
+import com.squareup.anvil.compiler.api.AnvilCompilationException
+import com.squareup.anvil.compiler.internal.reference.AnnotationReference
+import com.squareup.anvil.compiler.scope
+import org.jetbrains.kotlin.backend.common.extensions.IrPluginContext
+import org.jetbrains.kotlin.ir.expressions.IrConstructorCall
+import org.jetbrains.kotlin.ir.util.parentAsClass
+import org.jetbrains.kotlin.name.FqName
+import kotlin.LazyThreadSafetyMode.NONE
+
+internal class AnnotationReferenceIr(
+  val annotation: IrConstructorCall,
+  val classReference: ClassReferenceIr,
+  private val declaringClass: ClassReferenceIr?
+) {
+  val fqName: FqName
+    get() = classReference.fqName
+
+  private val context: IrPluginContext
+    get() = classReference.context
+
+  val scope: ClassReferenceIr by lazy(NONE) {
+    scopeOrNull ?: throw AnvilCompilationException(
+      element = annotation,
+      message = "Couldn't find scope for $fqName."
+    )
+  }
+
+  val scopeOrNull: ClassReferenceIr? by lazy(NONE) {
```

Comment:
Curious why you've chosen a property here, but not for declaringClass below. In the original implementation we use functions and hide the properties. Not a big deal, just curious :)

Reply:
Good question. I usually prefer properties for things like this where there aren't any inputs and it's not something you'd expect to change over the lifetime of the class. I defaulted to functions for declaringClass because of name conflicts but I think I can actually just adjust it to not be a property constructor and make the class more consistent.